cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2218,6 +2218,117 @@ TEST(Cast, ListToListOptionsPassthru) {
+  std::shared_ptr<Array> c, d;
+  c = ArrayFromJSON(int8(), "[1, 2]");
+  d = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto dest, StructArray::Make({c, d}, field_names2));
+  auto options = CastOptions::Safe(dest->type());
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      TypeError,
+      ::testing::HasSubstr("Type error: struct field names do not match: struct<a: int8, "
+                           "b: int8> struct<c: int8, d: int8>"),
+      Cast(src, options));
+}
+
+TEST(Cast, StructToDifferentSizeStruct) {
+  std::vector<std::string> field_names = {"a", "b"};
+  std::shared_ptr<Array> a, b;
+  a = ArrayFromJSON(int8(), "[1, 2]");
+  b = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto src, StructArray::Make({a, b}, field_names));
+
+  std::vector<std::string> field_names2 = {"a", "b", "c"};
+  std::shared_ptr<Array> a2, b2, c;
+  a2 = ArrayFromJSON(int8(), "[1, 2]");
+  b2 = ArrayFromJSON(int8(), "[3, 4]");
+  c = ArrayFromJSON(int8(), "[5, 6]");
+  ASSERT_OK_AND_ASSIGN(auto dest, StructArray::Make({a2, b2, c}, field_names2));
+  auto options = CastOptions::Safe(dest->type());
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      TypeError,
+      ::testing::HasSubstr("Type error: struct field sizes do not match: struct<a: int8, "
+                           "b: int8> struct<a: int8, b: int8, c: int8>"),
+      Cast(src, options));
+}
+
+TEST(Cast, StructToSameSizedButDifferentNullabilityStruct) {
+  // OK to go from not-nullable to nullable...
+  std::vector<std::shared_ptr<Field>> fields1 = {
+      std::make_shared<Field>("a", int8(), false),
+      std::make_shared<Field>("b", int8(), false)};
+  std::shared_ptr<Array> a1, b1;
+  a1 = ArrayFromJSON(int8(), "[1, 2]");
+  b1 = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto src1, StructArray::Make({a1, b1}, fields1));
+
+  std::vector<std::shared_ptr<Field>> fields2 = {
+      std::make_shared<Field>("a", int8(), true),
+      std::make_shared<Field>("b", int8(), true)};
+  std::shared_ptr<Array> a2, b2;
+  a2 = ArrayFromJSON(int8(), "[1, 2]");
+  b2 = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto dest1, StructArray::Make({a2, b2}, fields2));
+
+  CheckCast(src1, dest1);
+
+  // But not the other way around
+  std::vector<std::shared_ptr<Field>> fields3 = {
+      std::make_shared<Field>("a", int8(), true),
+      std::make_shared<Field>("b", int8(), true)};
+  std::shared_ptr<Array> a3, b3;
+  a3 = ArrayFromJSON(int8(), "[1, null]");
+  b3 = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto src2, StructArray::Make({a3, b3}, fields3));
+
+  std::vector<std::shared_ptr<Field>> fields4 = {
+      std::make_shared<Field>("a", int8(), false),
+      std::make_shared<Field>("b", int8(), false)};
+  std::shared_ptr<Array> a4, b4;
+  a4 = ArrayFromJSON(int8(), "[1, 2]");
+  b4 = ArrayFromJSON(int8(), "[3, 4]");
+  ASSERT_OK_AND_ASSIGN(auto dest2, StructArray::Make({a4, b4}, fields4));
+  auto options = CastOptions::Safe(dest2->type());
```

Comment:
FWIW, instead of constructing dest2, why not just use `arrow::struct_()` to directly construct the type?